data_scrapers/san_francisco.py

```diff
@@ -0,0 +1,222 @@
+    return meta_from_source
+
+def get_update_times(session: SocrataApi, resource_ids: Dict[str, str]) -> List:
+    """
+    Return a list of update times for all resources.
+    """
+    update_times = []
+    for v in resource_ids.values():
+        data = session.metadata(v)
+        update_times.append(data["dataUpdatedAt"])
+    return update_times
+
+def get_demographics(session: SocrataApi, resource_ids: Dict[str, str]) -> Dict:
+    """
+    Fetch cases by age, gender, race_eth. Fetch cases by transmission category
+    Returns the dictionary value for {"cases_totals": {}, "death_totals":{}}.
+    Note that SF does not provide death totals, so these datapoints will be -1.
+    To crete a DataFrame from the dictionary, run 'pd.DataFrame(get_demographics())'
+    """
+    # copy dictionary structure of global 'out' dictionary to local variable
+    demo_totals: Dict[str,Dict] = { "case_totals": dict(), "death_totals": dict() }
+    demo_totals["case_totals"]["gender"] = get_gender_table(session, resource_ids)
+    demo_totals["case_totals"]["age_group"] = get_age_table(session, resource_ids)
+    demo_totals["case_totals"]["transmission_cat"] = get_transmission_table(session, resource_ids)
+    demo_totals["case_totals"]["race_eth"] = get_race_eth_table(session, resource_ids)
+    return demo_totals
+
+def get_timeseries(session: SocrataApi, resource_ids: Dict[str, str]) -> Dict[str,List[Dict]]:
+    """
+    Returns the dictionary value for "series": {"cases":[], "deaths":[], "tests":[]}.
+    To create a DataFrame from this dictionary, run
+    'pd.DataFrame(get_timeseries())'
+    """
+    out_series: Dict[str, List[Dict]] = {"cases": [], "deaths": [
+    ], "tests": []}  # dictionary structure for time_series
+    out_series["cases"] = get_cases_series(session, resource_ids)
+    out_series["deaths"] = get_deaths_series(session, resource_ids)
+    out_series["tests"] = get_tests_series(session, resource_ids)
+    return out_series
+
+# Confirmed Cases and Deaths by Date and Transmission
+# Note that cumulative totals are not directly reported, we are summing over the daily reported numbers
+def get_cases_series(session : SocrataApi, resource_ids: Dict[str, str]) -> List[Dict]:
+    """Get cases timeseries json, sum over transmision cat by date"""
+    resource_id = resource_ids['cases_deaths_transmission']
+    params = {'case_disposition': 'Confirmed',
+            '$select': 'date,sum(case_count) as cases', '$group': 'date', '$order': 'date'}
+    data = session.resource(resource_id, params=params)
+    # convert date from ISO string to 'yyyy-mm-dd'. convert number strings to int.
+    # calculate daily cumulative
+    cases_series = []
+    cumul = 0
+    for entry in data:
+        entry["date"] = entry["date"][0:10]
+        entry["cases"] = int(entry["cases"])
+        cumul += entry["cases"]
+        entry["cumul_cases"] = cumul
+        cases_series.append(entry)
+    return cases_series
```

Comment:
I don’t think this is a blocker, but I just noticed we are creating time series with gaps here. e.g. I’m getting:

```
{
    "date": "2020-03-05",
    "cases": 2,
    "cumul_cases": 2
},
{
    "date": "2020-03-06",
    "cases": 5,
    "cumul_cases": 7
},
{
    "date": "2020-03-08",
    "cases": 6,
    "cumul_cases": 13
},
...
```

It looks like there are simply gaps in the source data (probably on days where there were 0 new cases/deaths), so it’s not something we’re doing *wrong* here.

If that’s not a serious issue, let’s not worry about it for this PR (and if we want to “fix” it, fix it as a separate PR).

This issue also applies to the other time series.

Reply:
Sounds good, I'll file this as a separate issue. Thanks for catching these.